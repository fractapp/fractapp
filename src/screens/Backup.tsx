import React, {useContext, useEffect} from 'react';
import {
  StyleSheet,
  View,
  ScrollView,
  Text,
  FlatList,
  Image,
  TouchableOpacity,
} from 'react-native';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import Dialog from 'storage/Dialog';
import BackupUtils from 'utils/backup';
import db from 'utils/db';
import backupUtil from 'utils/backup';

export const Backup = ({navigation, route}: {navigation: any; route: any}) => {
  const dialogContext = useContext(Dialog.Context);

  const onSuccess = () => {
    navigation.reset({
      index: 0,
      routes: [{name: 'Home'}],
    });
    dialogContext.dispatch(Dialog.close());
  };

  useEffect(() => {
    if (route.params?.isSuccess == null) {
      return;
    }

    const dir =
      route.params.type == BackupUtils.BackupType.File
        ? 'Downloads'
        : backupUtil.GoogleDriveFolder;

    dialogContext.dispatch(
      Dialog.open(
        'Success save wallet',
        `If you lose access to file then you will not be able to restore access to the wallet. File "${route.params?.fileName}.json" saved in "${dir}" directory`,
        onSuccess,
      ),
    );
  }, [route.params?.isSuccess]);

  const backupFile = async () => {
    const seed = await db.getSeed();
    await BackupUtils.backupFile(
      () =>
        navigation.navigate('WalletFileBackup', {
          seed: seed?.split(' '),
          type: BackupUtils.BackupType.File,
          callerScreen: 'Backup',
        }),
      () =>
        dialogContext.dispatch(
          Dialog.open(
            'Open settings',
            'If you want to save a file then open the application settings and give it access to the storage.',
            () => dialogContext.dispatch(Dialog.close()),
          ),
        ),
    );
  };
  const backupGoogleDrive = async () => {
    const seed = await db.getSeed();
    await BackupUtils.backupGoogleDrive(() =>
      navigation.navigate('WalletFileBackup', {
        seed: seed?.split(' '),
        type: BackupUtils.BackupType.GoogleDrive,
        callerScreen: 'Backup',
      }),
    );
  };
  const backupSeed = async () => {
    const seed = await db.getSeed();
    navigation.navigate('SaveSeed', {
      seed: seed?.split(' '),
      onSuccess: () =>
        navigation.reset({
          index: 0,
          routes: [{name: 'Home'}],
        }),
    });
  };

  const menuItems = [
    {
      img: (
        <Image
          source={require('assets/img/google-drive.png')}
          style={{width: 32, height: 32}}
        />
      ),
      title: 'Google disk',
      onClick: () => backupGoogleDrive(),
    },
    {
      img: <MaterialIcons name="folder-open" size={32} color="#888888" />,
      title: 'File to phone',
      onClick: () => backupFile(),
    },
    {
      img: (
        <MaterialCommunityIcons name="content-copy" size={32} color="#888888" />
      ),
      title: 'Backup seed',
      onClick: () => backupSeed(),
    },
  ];
  const renderItem = ({item}: {item: any}) => (
    <TouchableOpacity style={styles.menuItem} onPress={item.onClick}>
      {item.img}
      <Text style={styles.menuTitle}>{item.title}</Text>
    </TouchableOpacity>
  );

  return (
    <View style={styles.profile}>
      <ScrollView showsVerticalScrollIndicator={false} style={{width: '100%'}}>
        <View style={{alignItems: 'center', justifyContent: 'center'}}>
          <Text style={styles.description}>
            Save your wallet in a safe place. If you lose your wallet, you
            cannot restore access to it.
          </Text>
          <FlatList
            style={styles.menu}
            ItemSeparatorComponent={() => <View style={styles.dividingLine} />}
            data={menuItems}
            renderItem={renderItem}
            keyExtractor={(item) => item.title}
          />
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  profile: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  dividingLine: {
    alignSelf: 'center',
    width: '100%',
    borderBottomWidth: 1,
    borderBottomColor: '#DADADA',
  },
  description: {
    width: '90%',
    marginTop: 10,
    fontSize: 15,
    fontFamily: 'Roboto-Regular',
    color: '#888888',
  },
  menu: {
    flex: 1,
    width: '88%',
    marginTop: 20,
  },
  menuItem: {
    flexDirection: 'row',
    paddingTop: 10,
    paddingBottom: 10,
  },
  menuTitle: {
    marginLeft: 10,
    fontSize: 17,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    color: 'black',
    alignSelf: 'center',
  },
});
