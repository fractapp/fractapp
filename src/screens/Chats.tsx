import React, {useEffect} from 'react';
import {
  FlatList,
  StyleSheet,
  TouchableHighlight,
  TouchableOpacity,
  View,
} from 'react-native';
import {ChatShortInfo} from 'components/ChatShortInfo';
import {ChatInfo} from 'types/chatInfo';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { useDispatch, useSelector } from 'react-redux';
import ChatsStore from 'storage/Chats';
import UsersStore from 'storage/Users';
// @ts-ignore
import {MAIN_BOT_ID} from '@env';

/**
 * Chat list screen
 * @category Screens
 */
export const Chats = ({navigation}: {navigation: any}) => {
  const dispatch = useDispatch();
  const chatsState: ChatsStore.State = useSelector((state: any) => state.chats);
  const usersState: UsersStore.State = useSelector((state: any) => state.users);

  useEffect(() => {
    navigation.setOptions({
      headerRight: () => {
        return (
          <TouchableOpacity onPress={() => navigation.navigate('Search')}>
            <Ionicons name="add" size={32} color="#2AB2E2" />
          </TouchableOpacity>
        );
      },
    });
  }, []);

  const getChats = () => {
    const chats = Object.keys(chatsState.chatsInfo).map((key) => chatsState.chatsInfo[key])
      .filter((value) => value.lastMsgId != null)
      .sort((a, b) => {
          const msgOne = chatsState.chats[a.id]!.messages[a.lastMsgId!]!;
          const msgTwo = chatsState.chats[b.id]!.messages[b.lastMsgId!]!;
          if (msgOne.sender === MAIN_BOT_ID) {
            return -1;
          } else if (msgTwo.sender === MAIN_BOT_ID) {
            return 1;
          }

          return msgTwo.timestamp - msgOne.timestamp;
        }
      );

    return chats;
  };

  const renderItem = ({item}: {item: ChatInfo}) => {
    let user = usersState.users[item.id]!;

    if (!usersState.users[item.id]) {
      user = {
        isAddressOnly: false,
        title: 'Deleted',
        value: {
          id: item.id,
          name: 'Deleted',
          username: 'Deleted',
          avatarExt: '',
          lastUpdate: 0,
          addresses: null,
          isChatBot: false,
        },
      };
      dispatch(UsersStore.actions.setUsers([user]));
    }
    const msg = chatsState.chats[item.id]!.messages[item.lastMsgId!]!;

    return (
      <TouchableHighlight
        onPress={() => navigation.navigate('Chat', {chatId: item.id})}
        underlayColor="#f8f9fb">
        <ChatShortInfo
          notificationCount={item.notificationCount}
          message={msg.value}
          timestamp={msg.timestamp}
          user={user ?? {
            isAddressOnly: false,
            value: {
              id: '0',
              name: 'Deleted',
              username: '-',
              avatarExt: '',
              lastUpdate: 0,
              addresses: new Map(),
            },
          }}
        />
      </TouchableHighlight>
    );
  };

  return (
    <View style={styles.chats}>
      <FlatList
        ItemSeparatorComponent={() => <View style={styles.dividingLine} />}
        showsVerticalScrollIndicator={false}
        style={styles.list}
        data={getChats()}
        renderItem={renderItem}
        keyExtractor={(item, index) => String(index)}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  chats: {
    alignItems: 'center',
    justifyContent: 'center',
    flex: 1,
  },
  list: {
    flex: 1,
    width: '100%',
  },
  dividingLine: {
    marginLeft: 80,
    width: '100%',
    height: 1,
    backgroundColor: '#f5f5f5',
  },
});
