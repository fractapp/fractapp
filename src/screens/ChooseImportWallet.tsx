import React, {useContext, useEffect} from 'react';
import {StyleSheet, View, Text, FlatList} from 'react-native';
import GlobalStore from 'storage/Global';
import {WhiteButton} from 'components/WhiteButton';
import backup from 'utils/backup';
import googleUtil from 'utils/google';

/**
 * Choose import wallet file import screen
 * @category Screens
 */
export const ChooseImportWallet = ({
  route,
  navigation,
}: {
  route: any;
  navigation: any;
}) => {
  const globalContext = useContext(GlobalStore.Context);

  const type: backup.BackupType = route.params.type;
  const wallets: Array<string> = route.params.wallets;
  const ids: Array<string> = route.params?.ids;

  const open = (v: string) => {
    globalContext.dispatch(GlobalStore.setLoading(true));
    if (type === backup.BackupType.File) {
      backup
        .getFile(v)
        .then((file) => {
          navigation.navigate('WalletFileImport', {
            file: file,
          });
          setTimeout(
            () => globalContext.dispatch(GlobalStore.setLoading(false)),
            500,
          );
        })
        .catch((e) => {
          globalContext.dispatch(GlobalStore.setLoading(false));
        });
    } else if (type === backup.BackupType.GoogleDrive) {
      googleUtil
        .getFileBackup(v)
        .then((file) => {
          navigation.navigate('WalletFileImport', {
            file: file,
          });
          setTimeout(
            () => globalContext.dispatch(GlobalStore.setLoading(false)),
            500,
          );
        })
        .catch((e) => {
          globalContext.dispatch(GlobalStore.setLoading(false));
        });
    }
  };

  useEffect(() => {
    globalContext.dispatch(GlobalStore.setLoading(false));
  }, []);

  return (
    <View
      style={{
        flexDirection: 'column',
        flex: 1,
        alignItems: 'center',
      }}>
      <Text style={styles.title}>Choose wallet</Text>

      <FlatList
        style={{
          flex: 1,
          marginTop: 30,
          width: '90%',
        }}
        data={wallets}
        renderItem={({item}) => {
          return (
            <View style={{marginBottom: 20}}>
              <WhiteButton
                text={item.replace('.json', '')}
                height={50}
                onPress={() => {
                  if (type === backup.BackupType.File) {
                    open(wallets[0]);
                  } else if (type === backup.BackupType.GoogleDrive) {
                    open(ids[0]);
                  }
                }}
              />
            </View>
          );
        }}
        keyExtractor={(item, index) => item + index}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  noResultsText: {
    marginTop: 10,
    fontSize: 19,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: '#888888',
  },
  title: {
    marginTop: 80,
    fontSize: 25,
    fontFamily: 'Roboto-Regular',
    color: '#2AB2E2',
  },
  description: {
    textAlign: 'center',
    width: '90%',
    marginTop: 40,
    fontSize: 15,
    fontFamily: 'Roboto-Regular',
    color: '#888888',
  },
  newPassword: {
    marginTop: 30,
    width: '90%',
  },
  confirmPassword: {
    marginTop: 20,
    width: '90%',
  },
  error: {
    marginTop: 20,
    color: 'red',
    fontFamily: 'Roboto-Regular',
    fontSize: 15,
  },
});
