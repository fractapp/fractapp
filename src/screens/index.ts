export {ImportWallet} from './ImportWallet';
export {SaveSeed} from './SaveSeed';
export {SaveWallet} from './SaveWallet';
export {SettingWallet} from './SettingWallet';
export {Start} from './Start';
export {WalletDetails} from './WalletDetails';
export {WalletFileBackup} from './WalletFileBackup';
export {Wallets} from './Wallets';
export {Settings} from './Settings';
export {ConfirmSaveSeed} from './ConfirmSaveSeed';
export {ImportSeed} from './ImportSeed';
export {WalletFileImport} from './WalletFileImport';
export {GoogleDrivePicker} from './GoogleDrivePicker';
export {Receive} from './Receive';
export {Milestone} from './Milestone';
export {TransactionDetails} from './TransactionDetails';
export {SplashScreen} from './SplashScreen';
export {Backup} from './Backup';
export {NewPassCode} from './NewPassCode';
export {VerifyPassCode} from './VerifyPassCode';
export {EditProfile} from './EditProfile';
export {EditPhoneNumber} from './EditPhoneNumber';
export {SelectCountry} from './SelectCountry';
export {ConfirmCode} from './ConfirmCode';
export {EditEmail} from './EditEmail';
export {Navigation} from './Navigation';
export {Chats} from './Chats';
export {Chat} from './Chat';
export {Connecting} from './Connecting';
export {EditUsername} from './EditUsername';
export {EditName} from './EditName';
export {Search} from './Search';
export {SelectWallet} from './SelectWallet';
export {Send} from './Send';
export {EnterAmount} from './EnterAmount';
export {SendWithEnterAddress} from './SendWithEnterAddress';
