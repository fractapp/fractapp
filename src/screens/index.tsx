export { default as ImportWallet } from './ImportWallet';
export { default as SaveSeed } from './SaveSeed';

export { default as SaveWallet } from './SaveWallet';
export { default as SettingWallet } from './SettingWallet';
export { default as Start } from './Start';

export { default as WalletDetails } from './WalletDetails';

export { default as WalletEncryption } from './WalletEncryption';
export { default as Wallets } from './Wallets';
export { ConfirmSaveSeed } from './ConfirmSaveSeed'