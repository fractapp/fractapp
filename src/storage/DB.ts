import {Account} from 'models/account';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Keychain from 'react-native-keychain';
import {base64Encode, randomAsU8a} from '@polkadot/util-crypto';
import PasscodeUtil from 'utils/passcode';
import {Keyring} from '@polkadot/keyring';
import {u8aToHex} from '@polkadot/util';
import {Currency, getSymbol} from 'models/wallet';
import {Transaction} from 'models/transaction';
import {ChatInfo} from 'models/chatInfo';
import {AuthInfo} from 'models/authInfo';
import {MyProfile} from 'models/myProfile';
import BN from 'bn.js';
import {UserProfile} from 'models/profile';
import {Use} from 'react-native-svg';
/**
 * @namespace
 * @category storage
 */

namespace DB {
  const secureOption = {
    service: 'com.fractapp',
  };
  const biometryOption = {
    service: 'com.fractapp',
    accessControl: Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET,
  };

  const AsyncStorageKeys = {
    price: (currency: Currency) => `price_${currency}`,
    profile: 'profile',
    authInfo: 'auth_info',
    notificationCount: 'notification_count',
    accounts: 'accounts',
    contacts: 'contacts',
    users: 'users',
    accountInfo: (address: string) => `account_${address}`,
    transactions: (currency: Currency) => `transactions_${getSymbol(currency)}`,
    pendingTransactions: (currency: Currency) =>
      `pending_transactions_${getSymbol(currency)}`,
    chatByAddress: (address: string) => `chat_${address}`,
    chatsInfo: 'chats_info',
  };
  const SecureStorageKeys = {
    firebaseToken: 'firebase_token',
    authJWT: 'auth_jwt',
    seed: 'seed',
    salt: 'salt',
    passcodeHash: 'passcode_hash',
  };
  const PasscodeStorageKeys = {
    passcode: 'passcode',
  };

  export async function setSecureItem(key: string, value: string) {
    const result = await Keychain.setInternetCredentials(
      key,
      key,
      value,
      secureOption,
    );
    if (result == false) {
      throw `invalid set ${key}`;
    }
  }
  export async function getSecureItem(key: string): Promise<string | null> {
    const result = await Keychain.getInternetCredentials(key, secureOption);
    if (result === false) {
      return null;
    }

    return result.password;
  }

  export async function setAuthInfo(auth: AuthInfo) {
    await AsyncStorage.setItem(AsyncStorageKeys.authInfo, JSON.stringify(auth));
  }
  export async function getAuthInfo(): Promise<AuthInfo | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.authInfo);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setPrice(currency: Currency, value: number) {
    await AsyncStorage.setItem(AsyncStorageKeys.price(currency), String(value));
  }
  export async function getPrice(currency: Currency): Promise<number> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.price(currency));
    if (result == null) {
      return 0;
    }
    return Number(result);
  }

  export async function setProfile(profile: MyProfile) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.profile,
      JSON.stringify(profile),
    );
  }
  export async function getProfile(): Promise<MyProfile | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.profile);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setNotificationCount(count: number) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.notificationCount,
      String(count),
    );
  }
  export async function getNotificationCount(): Promise<number> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.notificationCount,
    );
    return Number(result);
  }

  export async function enablePasscode(passcode: string, isBiometry: boolean) {
    if (isBiometry) {
      const result = await Keychain.setInternetCredentials(
        PasscodeStorageKeys.passcode,
        PasscodeStorageKeys.passcode,
        passcode,
        biometryOption,
      );
      if (result === false) {
        throw 'invalid set passcode';
      }
    }

    const salt = base64Encode(randomAsU8a(32));
    const hash = PasscodeUtil.hash(passcode, salt);
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }

    try {
      await setSecureItem(SecureStorageKeys.salt, salt);
      await setSecureItem(SecureStorageKeys.passcodeHash, hash);
      authInfo.isPasscode = true;
      authInfo.isBiometry = isBiometry;
    } catch (e) {
      await disablePasscode();
    }

    await setAuthInfo(authInfo);
  }
  export async function disablePasscode() {
    let authInfo = await getAuthInfo();
    if (authInfo == null) {
      authInfo = new AuthInfo(false, false, false, false);
    }
    authInfo.isPasscode = false;
    authInfo.isBiometry = false;

    await setAuthInfo(authInfo);

    await Keychain.resetInternetCredentials(
      SecureStorageKeys.salt,
      secureOption,
    );
    await Keychain.resetInternetCredentials(
      SecureStorageKeys.passcodeHash,
      secureOption,
    );
  }

  export async function getSalt(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.salt);
  }
  export async function getPasscodeHash(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.passcodeHash);
  }
  export async function getPasscode(): Promise<string> {
    const result = await Keychain.getInternetCredentials(
      PasscodeStorageKeys.passcode,
      biometryOption,
    );
    if (result === false) {
      throw 'value by passcode key not found';
    }

    return result.password;
  }

  export async function createAccounts(seed: string) {
    let polkadotWallet = new Keyring({
      type: 'sr25519',
      ss58Format: 0,
    }).addFromUri(seed);
    let kusamaWallet = new Keyring({type: 'sr25519', ss58Format: 2}).addFromUri(
      seed,
    );
    let accountsInfo = new Array<Account>(
      new Account(
        'Polkadot wallet',
        polkadotWallet.address,
        u8aToHex(polkadotWallet.publicKey),
        Currency.Polkadot,
        0,
        new BN(0).toString(10),
      ),
      new Account(
        'Kusama wallet',
        kusamaWallet.address,
        u8aToHex(kusamaWallet.publicKey),
        Currency.Kusama,
        0,
        new BN(0).toString(10),
      ),
    );
    let accounts = new Array<string>();
    for (let i = 0; i < accountsInfo.length; i++) {
      const account = accountsInfo[i];
      accounts.push(account.address);
      await setAccountInfo(account);
    }
    await setAccounts(accounts);
    await setSecureItem(SecureStorageKeys.seed, seed);
  }
  export async function getSeed(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.seed);
  }

  export async function setAccounts(accounts: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accounts,
      JSON.stringify(accounts),
    );
  }
  export async function getAccounts(): Promise<Array<string> | null> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.accounts);
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setChatsInfo(chats: Map<string, ChatInfo>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatsInfo,
      JSON.stringify([...chats]),
    );
  }
  export async function getChatsInfo(): Promise<Map<string, ChatInfo>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.chatsInfo);

    if (result == null) {
      return new Map<string, ChatInfo>();
    }
    return new Map<string, ChatInfo>(JSON.parse(result));
  }

  export async function setChat(address: string, chats: Map<string, Currency>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.chatByAddress(address),
      JSON.stringify([...chats]),
    );
  }
  export async function getChat(
    address: string,
  ): Promise<Map<string, boolean>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.chatByAddress(address),
    );

    if (result == null) {
      return new Map<string, boolean>();
    }
    return new Map<string, boolean>(JSON.parse(result));
  }

  export async function setTx(currency: Currency, tx: Transaction) {
    const txs = await getTxs(currency);

    txs.set(tx.id, tx);
    await AsyncStorage.setItem(
      AsyncStorageKeys.transactions(currency),
      JSON.stringify([...txs]),
    );
  }
  export async function getTxs(
    currency: Currency,
  ): Promise<Map<string, Transaction>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.transactions(currency),
    );

    if (result == null) {
      return new Map<string, Transaction>();
    }
    return new Map<string, Transaction>(JSON.parse(result));
  }

  export async function setContacts(contacts: Array<UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.contacts,
      JSON.stringify(contacts),
    );
  }

  export async function getContacts(): Promise<Array<UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.contacts);

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setUsers(users: Map<string, UserProfile>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.users,
      JSON.stringify([...users]),
    );
  }

  export async function getUsers(): Promise<Map<string, UserProfile>> {
    const result = await AsyncStorage.getItem(AsyncStorageKeys.users);

    if (result == null) {
      return new Map<string, UserProfile>();
    }
    return new Map<string, UserProfile>(JSON.parse(result));
  }

  export async function setPendingTxs(currency: Currency, txs: Array<string>) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.pendingTransactions(currency),
      JSON.stringify(txs),
    );
  }

  export async function getPendingTxs(
    currency: Currency,
  ): Promise<Array<string>> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.pendingTransactions(currency),
    );

    if (result == null) {
      return [];
    }

    return JSON.parse(result);
  }

  export async function setAccountInfo(account: Account) {
    await AsyncStorage.setItem(
      AsyncStorageKeys.accountInfo(account.address),
      JSON.stringify(account),
    );
  }
  export async function getAccountInfo(
    address: string,
  ): Promise<Account | null> {
    const result = await AsyncStorage.getItem(
      AsyncStorageKeys.accountInfo(address),
    );
    if (result == null) {
      return null;
    }
    return JSON.parse(result);
  }

  export async function setFirebaseToken(token: string) {
    await setSecureItem(SecureStorageKeys.firebaseToken, token);
  }
  export async function getFirebaseToken(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.firebaseToken);
  }

  export async function setJWT(token: string) {
    await setSecureItem(SecureStorageKeys.authJWT, token);
  }
  export async function getJWT(): Promise<string | null> {
    return await getSecureItem(SecureStorageKeys.authJWT);
  }
}

export default DB;
