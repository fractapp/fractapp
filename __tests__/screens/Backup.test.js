import React, {useContext, useState} from 'react';
import renderer from 'react-test-renderer';
import {Backup} from 'screens/Backup';
import {fireEvent, render} from '@testing-library/react-native';
import DB from 'storage/DB';
import BackupUtils from 'utils/backup';
import GlobalStore from 'storage/Global';

jest.mock('storage/DB', () => ({
  getSeed: jest.fn(),
}));
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useContext: jest.fn(),
  Linking: jest.fn(() => ({
    openURL: jest.fn(),
  })),
}));
jest.mock('utils/backup', () => ({
  backupGoogleDrive: jest.fn(),
  BackupType: {
    GoogleDrive: 12,
    File: 13,
  },
  checkPermissions: jest.fn(),
}));
jest.mock('react-native-crypto', () => ({}));

useState.mockImplementation((init) => [init, jest.fn()]);

useContext.mockReturnValue({
  state: GlobalStore.initialState,
  dispatch: jest.fn(),
});

it('Test view', () => {
  const tree = renderer.create(<Backup navigation={null} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Test backup google drive', async () => {
  const navigate = jest.fn();
  const seed = 'seed seed';
  DB.getSeed.mockReturnValueOnce(seed);

  const component = render(<Backup navigation={{navigate: navigate}} />);
  await fireEvent.press(component.getByText('Google drive'));
  BackupUtils.backupGoogleDrive.mock.calls[0][0]();

  expect(navigate.mock.calls[0][0]).toMatchSnapshot();
  expect(navigate.mock.calls[0][1]).toMatchSnapshot();
});

it('Test backup file', async () => {
  const navigate = jest.fn();
  const seed = 'seed seed';
  DB.getSeed.mockReturnValueOnce(seed);

  const dispatch = jest.fn();
  useContext.mockReturnValueOnce({
    state: GlobalStore.initialState,
    dispatch: dispatch,
  });

  const component = render(<Backup navigation={{navigate: navigate}} />);
  await fireEvent.press(component.getByText('Encrypted file'));
  BackupUtils.checkPermissions.mock.calls[0][0]();
  BackupUtils.checkPermissions.mock.calls[0][1]();

  expect(navigate.mock.calls[0][0]).toMatchSnapshot();
  expect(navigate.mock.calls[0][1]).toMatchSnapshot();

  expect(dispatch).toMatchSnapshot();
});

it('Test seed', async () => {
  const navigate = jest.fn();
  const seed = 'seed seed';
  DB.getSeed.mockReturnValueOnce(seed);

  const component = render(<Backup navigation={{navigate: navigate}} />);
  await fireEvent.press(component.getByText('Backup seed'));

  expect(navigate.mock.calls[0][0]).toMatchSnapshot();
  expect(navigate.mock.calls[0][1]).toMatchSnapshot();
});
