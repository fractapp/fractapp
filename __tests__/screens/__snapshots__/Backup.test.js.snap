// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test backup google drive 1`] = `"WalletFileBackup"`;

exports[`Test backup google drive 2`] = `
Object {
  "isNewAccount": false,
  "seed": Array [
    "seed",
    "seed",
  ],
  "type": 12,
}
`;

exports[`Test seed 1`] = `"SaveSeed"`;

exports[`Test seed 2`] = `
Object {
  "isNewAccount": false,
  "seed": Array [
    "seed",
    "seed",
  ],
}
`;

exports[`Test view 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#888888",
        "fontFamily": "Roboto-Regular",
        "fontSize": 15,
        "marginTop": 10,
        "width": "90%",
      }
    }
  >
    [missing "en.backup.description" translation]
  </Text>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "img": <Image
            source={
              Object {
                "testUri": "../../../assets/img/google-drive.png",
              }
            }
            style={
              Object {
                "height": 32,
                "width": 32,
              }
            }
          />,
          "onClick": [Function],
          "title": "[missing \\"en.backup.google_drive_title\\" translation]",
        },
        Object {
          "img": <Icon
            allowFontScaling={false}
            color="#888888"
            name="content-copy"
            size={32}
          />,
          "onClick": [Function],
          "title": "[missing \\"en.backup.manually_title\\" translation]",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "marginTop": 20,
        "width": "88%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "opacity": 1,
              "paddingBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/img/google-drive.png",
              }
            }
            style={
              Object {
                "height": 32,
                "width": 32,
              }
            }
          />
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontFamily": "Roboto-Regular",
                "fontSize": 17,
                "fontStyle": "normal",
                "marginLeft": 10,
              }
            }
          >
            [missing "en.backup.google_drive_title" translation]
          </Text>
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "borderBottomColor": "#f5f5f5",
              "borderBottomWidth": 1,
              "width": "100%",
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "flexDirection": "row",
              "opacity": 1,
              "paddingBottom": 10,
              "paddingTop": 10,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#888888",
                  "fontSize": 32,
                },
                undefined,
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Û∞Üè
          </Text>
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "fontFamily": "Roboto-Regular",
                "fontSize": 17,
                "fontStyle": "normal",
                "marginLeft": 10,
              }
            }
          >
            [missing "en.backup.manually_title" translation]
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
